File: haskell/glue/src/Glue/Lib/IO/Read.hs
module Glue.Lib.IO.Read where

import Data.Text (pack)
import Glue.Eval (Eval, liftIO)
import Glue.IR (IR (..))

readLine :: IR Eval
readLine = NativeFunc readLineImpl

readLineImpl :: IR Eval -> Eval (IR Eval)
readLineImpl Void = do
    -- Accept Void as dummy argument for nullary call
    line <- liftIO getLine
    pure $ String (pack line)
readLineImpl _ = pure $ String ""

File: dart/glue/lib/src/lib/io/read.dart
import 'dart:io';
import 'package:glue/src/eval.dart';
import 'package:glue/src/ir.dart';

/// Read line function - reads a line from stdin
/// Mirrors Haskell Glue.Lib.IO.Read.readLine exactly
final Ir readLine = IrNativeFunc(readLineImpl);

/// Read line function implementation
/// Mirrors Haskell Glue.Lib.IO.Read.readLineImpl exactly
Eval<Ir> readLineImpl(Ir arg) {
  // Ignore the argument, always read a line
  return liftIO(stdin.readLineSync() ?? '').map((line) => IrString(line));
}
